bootloader/src/pyi_splash.c

```diff
@@ -42,6 +48,571 @@ typedef struct _ipc_message_head {
+    window_size.cx = ntohl(splash->wnd_width);
+    window_size.cy = ntohl(splash->wnd_height);
+
+    window_state->splash_bitmap = win32LoadBitmapHandle(splash);
+    window_state->splash_size = window_size;
+    window_state->splash_origin = win32WindowOrigin(window_state->splash_size);
+
+    if (window_state->splash_bitmap == NULL) {
+        _endthread();
+    }
+
+    /*
+     * Reallocate memory for splash as the bitmap color information in
+     * splash.bitmap is no longer needed since Windows now takes care of
+     * the resources for the bitmap.
+     */
+    tmp = splash;
+    splash = realloc(splash, splash->headerlen);
+
+    if (splash == NULL) {
+        FATALERROR("Cannot reallocate splash resources.\n");
+        splash = tmp;
+    }
+    window_state->splash_header = splash;
+
+    /* Create splash window */
+    splash_window = win32CreateWindow(window_state);
+
+    if (splash_window == NULL) {
+        FATALERROR("Splash Window Creation Failed.\n");
+        _endthread();
+    }
+    VS("SPLASH: Successful registration of all Microsoft Windows Handles.\n");
+
+    /* Show window and enter main loop */
+    ShowWindow(splash_window, SW_SHOW);
+
+    /* Main loop */
+    while (GetMessage(&msg, NULL, 0, 0)) {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    }
+
+    DeleteObject(window_state->splash_bitmap);
+    free(splash);
+    free(window_state);
+}
+
+/*
+ * Start an IPC server that waits for messages coming through the
+ * pipe and forwards them accordingly.
+ */
+static void
+win32IPCServer(void* param)
+{
+    HANDLE pipe_read_end;
+    DWORD num_of_bytes;
+    HWND splashWindow;
+
+    /* Use a simple protocol to communicate through the pipe
+     * ipc_message[0]   := event type
+     * ipc_message[1:5] := custom event data length
+     * ipc_message[>5]  := custom event data */
+    IPC_MESSAGE_HEAD msg_buffer = {0};
+    char *text_buffer = NULL;
+
+    pipe_read_end = (HANDLE) param;
+
+    /* Wait for another process to connect to the pipe. */
+    if (ConnectNamedPipe(pipe_read_end, NULL) == 0) {
+        FATAL_WINERROR("ConnectNamedPipe", "Failed to wait for connecting with pipe\n");
+        _endthread();
+    }
+
+    num_of_bytes = 0;
+
+    /* Loop that continuously reads data from the pipe, if any is present.
+     * ReadFile blocks if no data can be read, so this needs to run in a separate thread */
+    while (1) {
+        /* Read the head of the message */
+        ReadFile(pipe_read_end,
+                 &msg_buffer,
+                 sizeof(IPC_MESSAGE_HEAD),
+                 &num_of_bytes,
+                 NULL);
+
+        /* Client disconnected */
+        if (GetLastError() == ERROR_BROKEN_PIPE && num_of_bytes == 0) {
+            break;
+        }
+
+        /* If the message carries more data, load it into text_buffer */
+        if (msg_buffer.text_length != 0 && GetLastError() == ERROR_MORE_DATA) {
+            if (text_buffer != NULL) {
```

Comment:
Bug: If the user sends a empty message (meaning updating the text on the splash screen with an empty string) the old text inside the text_buffer is drawn again, since it is not freed.